autoload/lsp/diag.vim

```diff
@@ -171,55 +171,105 @@ def DiagsRefresh(bnr: number)
     diag_symbol = 'E>'
   endif
 
-  var signs: list<dict<any>> = []
   var diags: list<dict<any>> = diagsMap[bnr].sortedDiagnostics
-  for diag in diags
-    # TODO: prioritize most important severity if there are multiple diagnostics
-    # from the same line
-    var lnum = diag.range.start.line + 1
-    signs->add({id: 0, buffer: bnr, group: 'LSPDiag',
-				lnum: lnum,
-				name: DiagSevToSignName(diag.severity)})
-
-    try
-      if opt.lspOptions.highlightDiagInline
-        prop_add(diag.range.start.line + 1,
-                  util.GetLineByteFromPos(bnr, diag.range.start) + 1,
-                  {end_lnum: diag.range.end.line + 1,
-                    end_col: util.GetLineByteFromPos(bnr, diag.range.end) + 1,
-                    bufnr: bnr,
-                    type: DiagSevToInlineHLName(diag.severity)})
+  var diaglen = diags->len()
+  var ix: number
+
+  if &l:signcolumn != 'no'
+    var signs: list<dict<any>> = []
+    ix = 0
+    # Place line based diagnostics
+    while ix < diaglen
+      var diag = diags[ix]
+
+      # Prioritize most important severity when there are multiple diagnostics in
+      # the same line
+      while ix + 1 < diaglen
+          && diag.range.start.line == diags[ix + 1].range.start.line
+
+        if diag.severity > diags[ix + 1].severity
+          diag = diags[ix + 1]
+        endif
+
+        ix = ix + 1
+      endwhile
+
+      # increment for the iteration, place it as close to the top as possible,
+      # to avoid infinite loops, if "continue" is used.
+      ix = ix + 1
+
+      var lnum = diag.range.start.line + 1
+      if signs->len() == 0 || signs[-1].lnum != lnum
```

Comment:
I do not understand this logic here ```signs[-1].lnum != lnum``` when could this not be true?

Reply:
I think it was a relic from the attempt to have a one pass approach.